### PR TITLE
[ENHANCEMENT] Optimize regexps with multiple prefixes

### PR DESCRIPTION
**Pull request by bboreham**
For example `foo.*|bar.*|baz.*`. Instead of checking each one in turn, we build a map of prefixes, then check the smaller set that could match the string supplied.

Benchmarks (updated 2024-07-02):
```
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/model/labels
                                                       │  before.txt  │              after.txt               │
                                                       │    sec/op    │    sec/op      vs base               │
FastRegexMatcher/#00-8                                   44.89n ±  4%   43.62n ±   1%   -2.82% (p=0.002 n=6)
FastRegexMatcher/foo-8                                   51.32n ±  3%   51.75n ±   5%        ~ (p=0.193 n=6)
FastRegexMatcher/^foo-8                                  46.49n ±  1%   47.13n ±   2%        ~ (p=0.093 n=6)
FastRegexMatcher/(foo|bar)-8                             55.58n ±  2%   56.74n ±   1%        ~ (p=0.065 n=6)
FastRegexMatcher/foo.*-8                                 94.66n ±  4%   94.98n ±   1%        ~ (p=0.221 n=6)
FastRegexMatcher/.*foo-8                                 113.0n ±  3%   114.1n ±   1%        ~ (p=0.333 n=6)
FastRegexMatcher/^.*foo$-8                               112.3n ±  1%   114.8n ±   5%   +2.23% (p=0.002 n=6)
FastRegexMatcher/^.+foo$-8                               112.6n ±  0%   114.2n ±   0%   +1.51% (p=0.002 n=6)
FastRegexMatcher/.?-8                                    68.52n ±  5%   68.44n ±   1%        ~ (p=1.000 n=6)
FastRegexMatcher/.*-8                                    79.41n ± 10%   80.76n ±   3%        ~ (p=0.394 n=6)
FastRegexMatcher/.+-8                                    82.43n ±  5%   83.16n ±   1%        ~ (p=0.180 n=6)
FastRegexMatcher/foo.+-8                                 93.81n ±  1%   95.14n ±   1%   +1.42% (p=0.004 n=6)
FastRegexMatcher/.+foo-8                                 114.0n ±  4%   114.4n ±   4%        ~ (p=0.420 n=6)
FastRegexMatcher/foo_.+-8                                80.70n ±  2%   81.91n ±   1%   +1.49% (p=0.015 n=6)
FastRegexMatcher/foo_.*-8                                81.61n ±  3%   82.73n ±   3%        ~ (p=0.071 n=6)
FastRegexMatcher/.*foo.*-8                               192.7n ±  2%   199.5n ±   1%   +3.56% (p=0.002 n=6)
FastRegexMatcher/.+foo.+-8                               200.6n ±  2%   204.4n ±   2%        ~ (p=0.084 n=6)
FastRegexMatcher/(?s:.*)-8                               42.90n ±  2%   44.48n ±   1%   +3.68% (p=0.002 n=6)
FastRegexMatcher/(?s:.+)-8                               51.12n ±  5%   52.06n ±   2%        ~ (p=0.065 n=6)
FastRegexMatcher/(?s:^.*foo$)-8                          109.6n ±  0%   109.3n ±   1%        ~ (p=0.372 n=6)
FastRegexMatcher/(?i:foo)-8                              70.81n ±  2%   70.38n ±   2%        ~ (p=0.394 n=6)
FastRegexMatcher/(?i:(foo|bar))-8                        157.2n ±  3%   156.6n ±   6%        ~ (p=0.937 n=6)
FastRegexMatcher/(?i:(foo1|foo2|bar))-8                  283.8n ±  3%   285.7n ±   0%        ~ (p=0.394 n=6)
FastRegexMatcher/^(?i:foo|oo)|(bar)$-8                   710.2n ±  5%   722.5n ±   6%   +1.73% (p=0.041 n=6)
FastRegexMatcher/(?i:(foo1|foo2|aaa|bbb|ccc|ddd|e-8      798.0n ±  7%   800.5n ±   4%        ~ (p=0.818 n=6)
FastRegexMatcher/((.*)(bar|b|buzz)(.+)|foo)$-8           485.9n ±  2%   490.7n ±   1%   +0.99% (p=0.002 n=6)
FastRegexMatcher/^$-8                                    42.68n ±  3%   43.17n ±   1%        ~ (p=0.065 n=6)
FastRegexMatcher/(prometheus|api_prom)_api_v1_.+-8       186.8n ±  4%   187.2n ±   1%        ~ (p=0.190 n=6)
FastRegexMatcher/10\.0\.(1|2)\.+-8                       81.48n ±  1%   81.94n ±   3%        ~ (p=0.240 n=6)
FastRegexMatcher/10\.0\.(1|2).+-8                        80.74n ±  1%   82.16n ±   3%   +1.76% (p=0.002 n=6)
FastRegexMatcher/((fo(bar))|.+foo)-8                     204.9n ±  2%   205.8n ±   2%        ~ (p=0.732 n=6)
FastRegexMatcher/zQPbMkNO|NNSPdvMi|iWuuSoAl|qbvKM-8      203.5n ± 15%   215.4n ±   9%        ~ (p=0.093 n=6)
FastRegexMatcher/jyyfj00j0061|jyyfj00j0062|jyyfj9-8      204.2n ±  5%   213.8n ±   8%        ~ (p=0.180 n=6)
FastRegexMatcher/(?i:(zQPbMkNO|NNSPdvMi|iWuuSoAl|-8      794.8n ±  3%   803.5n ±   3%        ~ (p=0.240 n=6)
FastRegexMatcher/(?i:(AAAAAAAAAAAAAAAAAAAAAAAA|BB-8      356.9n ±  2%   357.9n ±   1%        ~ (p=1.000 n=6)
FastRegexMatcher/(?i:(.*zQPbMkNO|.*NNSPdvMi|.*iWu-8      7.481µ ±  1%   7.575µ ±   1%   +1.26% (p=0.004 n=6)
FastRegexMatcher/fo.?-8                                  93.49n ±  2%   95.07n ± 141%   +1.68% (p=0.024 n=6)
FastRegexMatcher/foo.?-8                                 94.03n ±  4%   94.62n ±   3%        ~ (p=0.310 n=6)
FastRegexMatcher/f.?o-8                                  81.32n ±  2%   80.81n ±   2%        ~ (p=0.937 n=6)
FastRegexMatcher/.*foo.?-8                               201.9n ±  2%   200.9n ±   0%   -0.45% (p=0.011 n=6)
FastRegexMatcher/.?foo.+-8                               192.2n ±  3%   189.9n ±   0%   -1.22% (p=0.002 n=6)
FastRegexMatcher/foo.?|bar-8                             147.5n ±  6%   156.1n ±   6%        ~ (p=0.106 n=6)
FastRegexMatcher/ſſs-8                                   51.20n ±  2%   51.90n ±   3%        ~ (p=0.132 n=6)
FastRegexMatcher/.*-.*-.*-.*-.*-8                        195.6n ±  2%   193.6n ±   2%        ~ (p=0.333 n=6)
FastRegexMatcher/(.+)-(.+)-(.+)-(.+)-(.+)-8              194.8n ±  1%   199.1n ±   4%   +2.21% (p=0.004 n=6)
FastRegexMatcher/((.*))(?i:f)((.*))o((.*))o((.*))-8      4.642µ ±  2%   4.616µ ±   1%        ~ (p=0.394 n=6)
FastRegexMatcher/((.*))f((.*))(?i:o)((.*))o((.*))-8      3.750µ ±  0%   3.802µ ±   2%   +1.40% (p=0.004 n=6)
FastRegexMatcher/(?i:(zQPbMkNO.*|NNSPdvMi.*|iWuuS#01-8   5.638µ ±  1%   1.040µ ±   1%  -81.55% (p=0.002 n=6)
FastRegexMatcher/(?i:(zQPbMkNO.*|NNSPdvMi.*|iWuuS-8      257.6n ±  1%   261.5n ±   3%   +1.49% (p=0.002 n=6)
geomean                                                  172.4n         168.4n          -2.32%
```

### CODE REVIEW

File: model/labels/regexp.go [hunk 1]
```diff
@@ -856,6 +884,8 @@ func optimizeEqualStringMatchers(input StringMatcher, threshold int) StringMatch
 		caseSensitive    bool
 		caseSensitiveSet bool
 		numValues        int
+		numPrefixes      int
+		minPrefixLength  = math.MaxInt
 	)
 
 	// Analyse the input StringMatcher to count the number of occurrences
```
Comment (pracucci) [inline model/labels/regexp.go:891]: [nit] I would rename `analyseCallback` to something like `analyseEqualMatcherCallback`

File: model/labels/regexp.go [hunk 2]
```diff
@@ -872,26 +902,44 @@ func optimizeEqualStringMatchers(input StringMatcher, threshold int) StringMatch
 		numValues++
 		return true
 	}
+	prefixCallback := func(matcher *literalPrefixStringMatcher) bool {
```
Comment (pracucci) [inline model/labels/regexp.go:905]: [nit] I would rename `prefixCallback ` to something like `analysePrefixMatcherCallback`

**Review by pracucci — COMMENTED** [review]
Good job Bryan! The logic looks solid to me. I left few minor comments, mostly to fix some comments and naming to match the new logic. I also suggest to improve a bit unit tests to add more cases with prefixes (e.g. alternation of prefixes with mixed case sensitivity).

Can you ping me once ready for a final review, please? 🙏

File: model/labels/regexp.go [hunk 3]
```diff
@@ -772,8 +782,17 @@ func (m *equalMultiStringMapMatcher) add(s string) {
 	m.values[s] = struct{}{}
 }
 
+func (m *equalMultiStringMapMatcher) addPrefix(p *literalPrefixStringMatcher) {
+	s := p.prefix[:m.minPrefixLen]
```
[the remark this answers is not in the document]
Comment (bboreham) [reply model/labels/regexp.go:786]: I'm pretty sure that code as it stands will panic if the string isn't long enough.  [Example](https://go.dev/play/p/yaAsM157aOu)

File: model/labels/regexp.go [hunk 4]
```diff
@@ -772,8 +782,17 @@ func (m *equalMultiStringMapMatcher) add(s string) {
...
+	if !m.caseSensitive {
+		s = strings.ToLower(s)
+	}
+
+	m.prefixes[s] = append(m.prefixes[s], p)
+}
+
 func (m *equalMultiStringMapMatcher) setMatches() []string {
-	if len(m.values) >= maxSetMatches {
+	if len(m.values) >= maxSetMatches || len(m.prefixes) > 0 {
```
[the remark this answers is not in the document]
Comment (bboreham) [reply model/labels/regexp.go:795]: That reminds me: 
Do you think 16 is still the right choice?

[the remark this answers is not in the document]
Comment (pracucci) [reply model/labels/regexp.go:795]: `maxSetMatches` is 256 and I don't remember how it was computed exactly (I remember we did few tests querying TSDB index at that time).

16 is the value of `minEqualMultiStringMatcherMapThreshold`. I computed the value using `BenchmarkOptimizeEqualStringMatchers` (see comment to `minEqualMultiStringMatcherMapThreshold`).

[the remark this answers is not in the document]
Comment (bboreham) [reply model/labels/regexp.go:795]: Sorry I see you are commenting specifically on `setMatches`.  I can add that case.

[the remark this answers is not in the document]
Comment (bboreham) [reply model/labels/regexp.go:795]: I checked: `setMatches` is only called from `optimizeAlternatingLiterals` which currently does not cover prefixes.
So it is not possible to create a case where `len(m.prefixes) > 0`.

**Comment by pracucci** [comment]
@bboreham I would like to move forward with this PR. With the aim to help, I opened a PR against your branch to address my own comments: https://github.com/bboreham/prometheus/pull/3

**Comment by bboreham** [comment]
Update on this: the extra benchmark Marco added shows that small sets of alternates slow down with my change, so more work is needed.

Meanwhile, I discovered another approach, to sort the alternates during regexp compilation https://github.com/grafana/regexp/pull/11, which works well when the strings have some similarity.

File: model/labels/regexp.go [hunk 5]
```diff
@@ -897,20 +938,24 @@ func (m trueMatcher) Matches(_ string) bool {
 	return true
 }
 
-// optimizeEqualStringMatchers optimize a specific case where all matchers are made by an
-// alternation (orStringMatcher) of strings checked for equality (equalStringMatcher). In
-// this specific case, when we have many strings to match against we can use a map instead
+// optimizeEqualOrPrefixStringMatchers optimize a specific case where all matchers are made by an
+// alternation (orStringMatcher) of strings checked for equality (equalStringMatcher) or
+// with a literal prefix (literalPrefixStringMatcher).
```
Comment (pracucci) [inline model/labels/regexp.go:943]: [nit]

```suggestion
// with a literal prefix (literalPrefixSensitiveStringMatcher or literalPrefixInsensitiveStringMatcher).
```

File: model/labels/regexp.go [hunk 6]
```diff
@@ -923,34 +968,55 @@ func optimizeEqualStringMatchers(input StringMatcher, threshold int) StringMatch
...
-	if numValues < threshold {
+	if (numValues + numPrefixes) < threshold {
 		return input
 	}
 
 	// Parse again the input StringMatcher to extract all values and storing them.
 	// We can skip the case sensitivity check because we've already checked it and
 	// if the code reach this point then it means all matchers have the same case sensitivity.
-	multiMatcher := newEqualMultiStringMatcher(caseSensitive, numValues)
+	multiMatcher := newEqualMultiStringMatcher(caseSensitive, numValues, numPrefixes, minPrefixLength)
```
Comment (pracucci) [inline model/labels/regexp.go:999]: If there's no prefix we pass `minPrefixLength=math.MaxInt`. Then, `equalMultiStringMapMatcher.Matches()` will check for `if m.minPrefixLen > 0`, which is true even if we're in the case "there no prefix". Likewise, the safety check done by `equalMultiStringMapMatcher.addPrefix()` is skrew up by this.

One way to simply fix this is:
- Initialise `minPrefixLength` to `0`
- Change `if len(prefix) < minPrefixLength {` to `if numPrefixes == 0 || (len(prefix) < minPrefixLength) {`

File: model/labels/regexp.go [hunk 7]
```diff
@@ -923,34 +968,55 @@ func optimizeEqualStringMatchers(input StringMatcher, threshold int) StringMatch
...
+
+		numPrefixes++
+		return true
+	}
+
+	if !findEqualOrPrefixStringMatchers(input, analyseEqualMatcherCallback, analysePrefixMatcherCallback) {
 		return input
 	}
 
 	// If the number of values found is less than the threshold, then we should skip the optimization.
```
Comment (pracucci) [inline model/labels/regexp.go:991]: [nit]

```suggestion
	// If the number of values and prefixes found is less than the threshold, then we should skip the optimization.
```

File: model/labels/regexp.go [hunk 8]
```diff
@@ -923,34 +968,55 @@ func optimizeEqualStringMatchers(input StringMatcher, threshold int) StringMatch
...
 
 	return multiMatcher
 }
 
-// findEqualStringMatchers analyze the input StringMatcher and calls the callback for each
-// equalStringMatcher found. Returns true if and only if the input StringMatcher is *only*
-// composed by an alternation of equalStringMatcher.
-func findEqualStringMatchers(input StringMatcher, callback func(matcher *equalStringMatcher) bool) bool {
+// findEqualOrPrefixStringMatchers analyze the input StringMatcher and calls the equalMatcherCallback for each
+// equalStringMatcher found, and prefixMatcherCallback for each literalPrefixStringMatcher found.
```
Comment (pracucci) [inline model/labels/regexp.go:1015]: [nit]

```suggestion
// equalStringMatcher found, and prefixMatcherCallback for each literalPrefixSensitiveStringMatcher and literalPrefixInsensitiveStringMatcher found.
```

File: model/labels/regexp.go [hunk 9]
```diff
@@ -923,34 +968,55 @@ func optimizeEqualStringMatchers(input StringMatcher, threshold int) StringMatch
...
+//
+// Returns true if and only if the input StringMatcher is *only* composed by an alternation of equalStringMatcher and/or
+// literalPrefixStringMatcher. Returns false if prefixMatcherCallback is nil and a literalPrefixStringMatcher is encountered.
```
Comment (pracucci) [inline model/labels/regexp.go:1018]: [nit]
```suggestion
// literal prefix matcher. Returns false if prefixMatcherCallback is nil and a literal prefix matcher is encountered.
```

**Review by pracucci — APPROVED** [review]
Very nice work, LGTM! I've also locally run some fuzzy tests and they couldn't spot any issue. I left few minor comments I would like you to consider before merging, thanks!

**Review by pracucci — APPROVED** [review]
LGTM, thanks for addressing my feedback!

**Pull request merged**